### ISSUE TITLE
HTTPResponse.closed consistency

### ISSUE THREAD

**Issue by nateprewitt** [shazow/urllib3#977]
I spent a while yesterday trying to figure out why responses were hitting logic requiring the file to be closed while still evaluating to `closed=False`. There's currently an inconsistency in how urllib3 handles closed responses and part of that is due to #929.

First off, I guess we need to clarify what "closed" means for the HTTPResponse. Does it mean the content has all been read, or that we've closed the file pointer? Perhaps it's both and we should be marking the response closed when all the content is read?

Background
--------------
Right now, we have two separate variables `isclosed` (Python 2&3) and `closed` (Python 3) on Python's `HTTPResponse` object. They both behave differently in what determines their boolean value. `isclosed` is defined by the `HTTPResponse`, whether that be in `httplib` (Python 2) or `http.client` (Python 3), while `closed` is defined by the `io` module that `http.client.HTTPResponse` inherits from in Python 3.

`isclosed` checks if the file pointer's content is `None` while the `io` module's `closed` attr only evaluates to True when set. This means we can read all of the bytes off the HTTPResponse object but still have an "open" file pointer with a length of 0 when using `io`'s `closed` attr. This won't change until `close()` is called on the file pointer,  but since Python 2 doesn't have a `closed` attr it drops through the first check in urllib3's `closed` method an evaluates as currently expected. Perhaps it's the Python 2 logic that's flawed though, and we shouldn't be leaving cleanup to the GC (assuming the response is ever dereferenced).

I put together 69018f7 to address the issue but didn't open the PR immediately because of the GC question and a side effect with urllib3's `close` method. Moving `isclosed` to a higher preference than `closed` causes `close` to no longer actually close the file pointer as it previously did in Python3. We can obviously implement something to address this but I wanted to get another pair of eyes on this before moving forward.

Example
-----------
This is a quick demo for Python 3 with the current master branch, as is the test included in the PR:
```python
from urllib3.connectionpool import HTTPConnectionPool
from urllib3.util.response import is_fp_closed

conn = HTTPConnectionPool('stackoverflow.com', maxsize=10)
get_response = conn.request('GET', url='/', preload_content=False)
data = [chunk for chunk in get_response.stream(1024)]
print(is_fp_closed(get_response._fp), get_response.closed)
```

**Comment by Lukasa** [comment]
Response closure is not a well defined property in httplib, or in urllib3. This is primarily because closure of a *response* is an ill-defined issue semantically: responses don't get closed, they just finish. *Connections*, they get closed.

So let's address some confusion. Responses are definitionally closed if their underlying fp is closed, but in the case of urllib3 it's more complete to say that a Response is closed when it behaves like a closed file-like-object. That means a response is closed if any subsequent call to `read()` would return an empty byte string, or if `close()` has been called on it (which should hopefully cause the first behaviour). Note that all of this is *deliberately* unrelated to the underlying fp, and in fact calling `close` on a response should only sometimes close the underlying fp. Frequently the response doesn't even *have* the underlying fp in order to close it!

What's not clear to me yet is exactly what issue you're bumping into. Can you clarify with a more explicit description of what problem is being encountered under what circumstances?

**Comment by nateprewitt** [comment]
I agree that the closure of a response doesn't really make sense, but the line was blurred a bit between content and connection when it began inheriting from a file-like super class in Python 3.

>So let's address some confusion. Responses are definitionally closed if their underlying fp is closed, but in the case of urllib3 it's more complete to say that a Response is closed when it behaves like a closed file-like-object. That means a response is closed if any subsequent call to read() would return an empty byte string, or if close() has been called on it (which should hopefully cause the first behaviour). 

Great, I think this agrees then that it makes more sense to use `isclosed` to determine response "closure" than the `closed` attr inherited from `io`. httplib/http.client implemented it for the reasons you specified.

>What's not clear to me yet is exactly what issue you're bumping into. Can you clarify with a more explicit description of what problem is being encountered under what circumstances?

**(This works for both stream and a single read() call, but I'll use stream in this example)**
So the case here is, I have a response. I want to read 1024 bytes of the response, append them to different list depending on their content, and continue streaming until there's no data left. If I try to use something like:

```python
while not response.closed:
    data = response.stream(1024)
    if data is x:
        list1.append(data)
    if data is y:
        list2.append(data)
```

This will continue to return a generator of `b''` forever after it has reached the end of the content. In it's current state `closed` on urllib3's HTTPResponse, in Python 3, will never change unless I write my own logic to call `close()` on it. This is the same bug we were solving in #929, just in a different check. I would say that `closed` currently has no useful meaning for Python 3. If `closed` is supposed to represent that there is no more data to be read, then we should put `isclosed` as the higher preference.

We'll still need a way to address `urllib3.response.HTTPResponse`'s `close()` no longer being able to close the `fp` though. `isclosed` will evaluate to `True` since the content is all read, but the underlying `fp` if it exists, may still be open. We are currently trying to close the `fp`, so 69018f7 alone would be a loss of functionality.

This is all based on my interpretation that `closed` is the only usable public equivalent of `is_fp_closed` so that we're not having to touch private variables, but perhaps that's a naive interpretation.

**Comment by nateprewitt** [comment]
@Lukasa, I'm redacting my last comment for a moment while I investigate a couple things I think I misinterpreted.

**Comment by nateprewitt** [comment]
I agree that the closure of a response doesn't really make sense, but the line was blurred a bit between content and connection when it began inheriting from a file-like super class in Python 3.

>So let's address some confusion. Responses are definitionally closed if their underlying fp is closed, but in the case of urllib3 it's more complete to say that a Response is closed when it behaves like a closed file-like-object. That means a response is closed if any subsequent call to read() would return an empty byte string, or if close() has been called on it (which should hopefully cause the first behaviour). 

Great, I think this agrees then that it makes more sense to use `isclosed` to determine response "closure" than the `closed` attr inherited from `io`. httplib/http.client implemented it for the reasons you specified.

>What's not clear to me yet is exactly what issue you're bumping into. Can you clarify with a more explicit description of what problem is being encountered under what circumstances?

I redacted my original response to this because the original example that I hit turns out to be similar but unrelated behaviour. Let me try to break down what I'm seeing a bit clearer.

Here's the three main ways I see to consume data:

**Stream** (for all integer amounts)
```python
from urllib3.connectionpool import HTTPConnectionPool
from urllib3.util.response import is_fp_closed

conn = HTTPConnectionPool('stackoverflow.com', maxsize=10)
get_response = conn.request('GET', url='/', preload_content=False)
data = [chunk for chunk in get_response.stream(1024)]
```
**Python 2:** `get_response.closed == True`
**Python 3:** `get_response.closed == False`

---
**Read(amt)**
```python
from urllib3.connectionpool import HTTPConnectionPool
from urllib3.util.response import is_fp_closed

conn = HTTPConnectionPool('stackoverflow.com', maxsize=10)
get_response = conn.request('GET', url='/', preload_content=False)
while get_response.closed:
    data = get_response.read(1024)
```
**Python 2:** `get_response.closed == True`
**Python 3:** `get_response.closed == True`

---
**Read()**
```python
from urllib3.connectionpool import HTTPConnectionPool
from urllib3.util.response import is_fp_closed

conn = HTTPConnectionPool('stackoverflow.com', maxsize=10)
get_response = conn.request('GET', url='/', preload_content=False)
data = get_response.read()
```
**Python 2:** `get_response.closed == True`
**Python 3:** `get_response.closed == False`

---
The only reason `read(amt)` (and now `stream(None)` post-#929) works is that we currently [force the connection closed](https://github.com/shazow/urllib3/blob/master/urllib3/response.py#L389) which sets `closed`. Calling just `read()` or `stream(amt)` though will consume all of the content, identically to `read(amt)`, without setting the `closed` flag on the `httplib`/`http.client` `HTTPResponse`. This means in Python 3, urllib3's `closed` will continue to be `False` despite being `True` in all other cases.

If `closed` is supposed to represent that there is no more data to be read, then we should put `isclosed` as the higher preference so this check works consistently across versions. Alternatively, we could also determine when a full read is completed and force the fp closed there so things are consistent. The reason we're currently forcing the fp shut on streams though seems to be coincidental and unrelated to any of these issues.

If we choose to prefer `isclosed`, we'll still need a way to address `urllib3.response.HTTPResponse`'s `close()` no longer being able to close the `fp` though. `isclosed` will evaluate to `True` since the content is all read, but the underlying `fp` if it exists, may still be open. We are currently trying to close the `fp`, so 69018f7 alone would be a loss of functionality.

This is all based on my interpretation that `closed` is the only usable public equivalent of `is_fp_closed` so that we're not having to touch private variables, but perhaps that's a naive interpretation.

**Comment by nateprewitt** [comment]
Ok @Lukasa, sorry these comments are so insanely verbose.

After a further dive this morning, this is definitely a conflicting definitions of "closed" between an `HTTPResponse` and an `io` object. `http.client.HTTPResponse` is conforming to the `io` API which requires explicit closure of the object outside of `read`.

`io` objects will never be `closed` until we either manually call `close()` on them, or dereference them so `__exit__` fires on cleanup. Neither of these situations are terribly useful for us. `http.client.HTTPResponse` dereferences the `fp` as soon as we finish reading with `_close_conn`, closing the underlying `fp` object. This means my initial concerns of changing how `urllib3.HTTPResponse.close()` were unfounded. We can just change the precedence and things will work consistently.

Example
---------

```python
from urllib3.connectionpool import HTTPConnectionPool

conn = HTTPConnectionPool('stackoverflow.com', maxsize=10)
response = conn.request('GET', url='/', preload_content=False)
httpresponse = response._fp

# we have access to the socket
assert httpresponse.fp is not None
fp_refer = httpresponse.fp
data = httpresponse.read()

# socket is consumed and _close_conn closed it
assert httpresponse.fp is None
assert fp_refer.closed

# HTTPResponse still aren't "closed"
assert not httpresponse.closed
httpresponse.close()

# now we are
assert httpresponse.closed
```

**Comment by Lukasa** [comment]
Does any of the urllib3 codebase rely on `response.closed`? Or do we just provide it as a helper for users?

**Comment by Lukasa** [comment]
A quick grep suggests that it's just a helper, though I'm prepared to be told otherwise. If it's just a helper, I'm pretty happy with you writing a test that demonstrates the problem and then providing a fix.

**Comment by nateprewitt** [comment]
Yeah, it's primarily a helper. The only location that it's currently used in the codebase is in [close](https://github.com/shazow/urllib3/blob/master/urllib3/response.py#L475-L476), which was my original concern. Now that we know this won't have any negative side effects on it though, I'll open the PR. `isclosed` is hardcoded in the places it actually matters too.

**Comment by Lukasa** [comment]
Resolved in #978.
